### ISSUE TITLE
CRDs are left after og and subscription are removed

### ISSUE THREAD

**Issue by gyliu513** [operator-framework/operator-lifecycle-manager#1326]
I used subscription deployed an operator, but after the subscription was deleted, the CRD for the operator was not deleted, is this a bug or design behavior for OLM? Thanks.

**Comment by dinhxuanvu** [comment]
Hey Guang,

This is the expected behavior as CRDs will not get deleted by OLM even when the operator is deleted. Also, when you say you delete an operator, you will need to delete both ‘Subscription’ and ‘CSV’ as well. Just FYI.

Thanks,
Vu

**Comment by gyliu513** [comment]
@dinhxuanvu Yes, I deleted both ‘Subscription’ and ‘CSV’, but why CRDs were not deleted? Do you have some document clarifying that this is design behavior? Thanks.

**Comment by dinhxuanvu** [comment]
Hey Guang,
We haven't had a documentation to capture this part yet (working on it). It is expected behavior for operator to clean up its own custom resources.

**Comment by gyliu513** [comment]
/reopen

@dinhxuanvu Here is not CR but CRD for the operator. I think the operator have no chance to delete the CRD, the OLM should handle this case.

**Comment by ecordell** [comment]
OLM is very conservative in what it deletes. Removing an operator does not necessarily mean that you want to lose all related data (you may be attempting to switch to another operator managing the same APIs, or you may have other processes that rely on reading the instances of the CRD). 

Today, if that is what you want, it’s not too onerous to delete the CRDs you no longer need. 

On the radar is adding a configurable finalizer which can optionally clean up for you, but it’s not obvious that this should happen in all cases. Does that sound like would address your needs?

**Comment by gyliu513** [comment]
> On the radar is adding a configurable finalizer which can optionally clean up for you

Thanks @ecordell this will help, but I think OLM does not support this yet?

**Comment by awgreene** [comment]
>Thanks @ecordell this will help, but I think OLM does not support this yet?

Hello @gyliu513, that is correct.

**Comment by gyliu513** [comment]
Thanks @awgreene , then I think we still need to keep this open. It would be great if the OLM can provide some option to enable end user can delete the CRDs for the operator. :-)

**Comment by flickerfly** [comment]
The delete is a feature of kubernetes already since it is a Kubernetes resource. Can't the end user just run `kubectl delete crds\crdname`? Given the various situations, I think the caution here to not delete automatically is very warranted.

**Comment by gyliu513** [comment]
Thanks @flickerfly , can you please share a use case when CRD should not be deleted when the operator was deleted?

I'm with you that we should delete CRD with caution, but my thinking is that after operator was deleted, all its related resources should be deleted including CRD as well.

**Comment by dinhxuanvu** [comment]
I can see the use case for CRD that involves storage and database. You probably would like to defer deleting CRD so you properly backup the storage/db if needed or even delete those storage/db to avoid leaking sensitive data and such.

**Comment by gyliu513** [comment]
@dinhxuanvu one question for this: If the operator was deleted and only left some CRD, then how to back up the storage/db without a operator controller?

Can you please show a more detailed use case? Thanks.

**Comment by dinhxuanvu** [comment]
I haven't tried out this operator specifically but I can see this case of AWS S3 operator (https://operatorhub.io/operator/awss3-operator-registry) which has a CRD (OBC) that can provision a S3 Bucket in AWS. That storage is external to the cluster where the operator is installed. When you delete the operator, you probably want to handle the existing S3 Buckets so they won't get orphaned and just hanging around in your AWS account which potentially cost you money.

**Comment by gyliu513** [comment]
A very good example, thanks @dinhxuanvu !

I agree that we can delete the CRD by the cluster admin manually, this is a good option. But the question is, in a large cluster, there maybe operators coming and go, and thus there will be lot of garbage CRDs exist in the cluster, and the admin need to check which CRD should be deleted or not. 

And end user may also have some concern as this: The OLM can help me install the CRD, why cannot uninstall/remove the CRD via OLM?

So I was still wondering can OLM add a feature to support delete CRD via OLM? As my comments at https://github.com/operator-framework/operator-lifecycle-manager/issues/1326#issuecomment-591960396

**Comment by dinhxuanvu** [comment]
I think it is a good feature for OLM to have that allow operator authors to specify how operator resources to be handled/deleted during its lifecycle including when it is deleted. I expect this feature to be added to OLM in the near future.

**Comment by njhale** [comment]
@gyliu513 I updated the title to more closely match what I think the feature discussed on this thread may look like. Please LMK if it's incorrect.

**Comment by gyliu513** [comment]
Looks good, thanks @njhale !

**Comment by dlandis** [comment]
> I expect this feature to be added to OLM in the near future.

Was this feature ever added, or no?

**Comment by dinhxuanvu** [comment]
This issue is stale. Closed. Please feel free to reopen if needed.
